### PR TITLE
fix #19175 How should addTime be used from a test?

### LINKED ISSUE #19175: How should addTime be used from a test?
When a testWidgets test times out, the following message is printed:

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TimeoutException was thrown running a test:
TimeoutException after 0:00:02.100000: The test exceeded the timeout. It may have hung.
Consider using "addTime" to increase the timeout before expensive operations.
```

It's unclear how "addTime" should be called from a test case. Maybe:

```dart
    final TestWidgetsFlutterBinding binding = tester.binding;
    if (binding is AutomatedTestWidgetsFlutterBinding)
      binding.addTime(const Duration(seconds: 30));
```

Seems complicated. The expected way should be documented.

It is also surprising to me that setting the `timeout` property on `testWidgets` alone doesn't increase the actual test timeout and I have to do the `addTime` hack in addition to setting the `timeout` property. This probably also needs a documentation explanation how the two relate to each other. 

/cc @Hixie

### PR DESCRIPTION
**Pull request by truongsinh**
fix #19175 How should addTime be used from a test?

### CODE REVIEW

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 1]
```diff
@@ -228,6 +231,20 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
...
   }
 
+  /// See [AutomatedTestWidgetsFlutterBinding.addTime]
+  ///
+  /// This is the shorthand for
+  /// ```dart
+  ///     final AutomatedTestWidgetsFlutterBinding binding = tester.binding;
+  ///     binding.addTime(const Duration(seconds: ));
+  /// ```
+  void addTime(Duration duration) {
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart]: I think in https://github.com/flutter/flutter/issues/19175#issuecomment-403607807 it was suggested to add this to `TestWidgetsFlutterBinding` instead as an empty implementation and then add an `@override` annotation to the existing impl in `AutomatedTestWidgetsFlutterBinding`.

**Review by goderbauer — COMMENTED** [review]
Thanks for the contribution.

To fully fix https://github.com/flutter/flutter/issues/19175 you'd also have to adress the follwong:

> * Add Documentation to the timeout property of testWidgets explaining its relationship to addTime.
> * Improve the timeout message to include more information of how to use addTime and how it relates to other test-related timeouts.

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter_test/lib/src/widget_tester.dart]: The `TestWidgetsFlutterBinding` has some other timeout related stuff as well (e.g. `defaultTestTimeout`). To keep everything at the same place, `addTime` should go here as well. Plus, that doesn't make the call side any more complicated. Both of your examples are almost equally concise. 

Additionally, moving it over there avoids the awkward type check you currently have in the implementation.

**Comment by goderbauer** [comment]
What I meant by my previous comment:

* Documentation should be added to the timeout property of testWidget (see [1]) to explain how increasing that timeout right there relates to adding more time with `addTime` - and also to educate people who look there about the presence of `addTime`

* https://github.com/flutter/flutter/pull/25202/files#diff-1fc62c212be6fd3a5e10c523ea69cb28R836 still doesn't mention testWidget's timeout property.

[1] https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/widget_tester.dart#L35

**Comment by goderbauer** [comment]
@truongsinh Are you still planning to come back to this?

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 1 above)
Comment (truongsinh) [reply packages/flutter_test/lib/src/widget_tester.dart]: Frankly, I am not totally convinced that "examples are almost equally concise", and do not necessarily agree that timeout related stuff should be in `TestWidgetsFlutterBinding`. However, a compromise right now is better than nothing, so I will continue anyway.

(The discussion on the conciseness/aesthetic of the sample code can continue later)

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 2]
```diff
@@ -60,6 +62,7 @@ typedef WidgetTesterCallback = Future<void> Function(WidgetTester widgetTester);
 ///
 /// ```dart
 /// testWidgets('MyWidget', (WidgetTester tester) async {
+///   tester.binding.addTime(const Duration(seconds: 3));
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart]: We shouldn't show this in the example. Most people will never need this.

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 3]
```diff
@@ -260,6 +263,16 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
     });
   }
 
+  /// See [AutomatedTestWidgetsFlutterBinding.addTime]
+  ///
+  /// This is the shorthand for
+  /// ```dart
+  ///     tester.binding.addTime(const Duration(seconds: ));
+  /// ```
+  void addTime(Duration duration) {
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart]: Can you remove this all-together?

**Review by goderbauer — CHANGES_REQUESTED** [review]

**Comment by goderbauer** [comment]
There are also various analyzer errors that need to be fixed: https://cirrus-ci.com/task/5132347000225792

And this point is still unaddressed:

> Add Documentation to the timeout property of testWidgets explaining its relationship to addTime.

File: packages/flutter_test/lib/src/binding.dart [hunk 4]
```diff
@@ -125,6 +125,15 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
   @protected
   bool get disableShadows => false;
 
+  /// Abstract method, implemented in [AutomatedTestWidgetsFlutterBinding] only.
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/binding.dart]: Can you replace this paragraph with `Increase the timeout for the current test by the given duration.`? The doc comment shouldn't explain how something is implemented, just what it is used for.

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 3 above)
Comment (truongsinh) [reply packages/flutter_test/lib/src/widget_tester.dart]: will do

**Comment by truongsinh** [comment]
Thanks for proactively reviewing this, I marked this as WIP as there are several things to be resolved, though reviewing during MR is welcome :D

**Comment by truongsinh** [comment]
@goderbauer does the latest commit https://github.com/flutter/flutter/pull/25202/commits/ca36fe11bf9954356359cc37f6ba9de8c14b8e64  address your point ?

> Add Documentation to the timeout property of testWidgets explaining its relationship to addTime.

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 5]
```diff
@@ -48,13 +48,25 @@ typedef WidgetTesterCallback = Future<void> Function(WidgetTester widgetTester);
...
+/// which defaults to [TestWidgetsFlutterBinding.defaultTestTimeout].
 ///
 /// This function uses the [test] function in the test package to
 /// register the given callback as a test. The callback, when run,
 /// will be given a new instance of [WidgetTester]. The [find] object
 /// provides convenient widget [Finder]s for use with the
 /// [WidgetTester].
+/// 
+/// See also:
+///  * [AutomatedTestWidgetsFlutterBinding.addTime] to learn more about
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart]: nit: please add a blank line after "See also:"

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 6]
```diff
@@ -48,13 +48,25 @@ typedef WidgetTesterCallback = Future<void> Function(WidgetTester widgetTester);
...
+/// timeout and how to manually increase timeout
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart]: timeout -> timeouts.

Also, please end with a period `.`

File: packages/flutter_test/lib/src/binding.dart [hunk 7]
```diff
@@ -125,6 +125,14 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
   @protected
   bool get disableShadows => false;
 
+  /// Increase the timeout for the current test by the given duration.
+  void addTime(Duration duration)
+    // noop
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/binding.dart]: I would replace this entire paragraph with:

```dart
// Noop, see [AutomatedTestWidgetsFlutterBinding. addTime] for an actual implementation.
```

**Review by goderbauer — COMMENTED** [review]
This is looking good. Just a couple of phrasing nits.

**Comment by truongsinh** [comment]
Thanks goderbauer, really appreciate your throughout review and patience with me :D

**Review by goderbauer — APPROVED** [review]
LGTM

Thanks for the contribution!

**Comment by goderbauer** [comment]
I'll submit this once the tree is green again.

**Pull request merged**